Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -3114,17 +3116,41 @@ def _inject_eval(self, node, func_name):
                     node.pos, self.current_scope_node(), lenv))
         return node
 
+    def _check_inside_class(self, def_node):
+        return isinstance(def_node, Nodes.DefNode) and def_node.args and len(self.env_stack) >= 2
+        
+
+    def _get_current_class_and_scope(self):
+        return self.env_stack[-2]
+    
+    def _inject_class(self, node):
+        # bare __class__ reference inside function
+        lenv = self.current_env()
+        entry = lenv.lookup_here(u"__class__")
+        if entry:
+            # Already have a __class__ local
+            return
+
+        def_node = self.current_scope_node()
+        if not self._check_inside_class(def_node):
+            return
+        
+        class_node, class_scope = self._get_current_class_and_scope()
+        if class_scope.is_py_class_scope:
+            def_node.requires_classobj = True
+            class_node.class_cell.is_active = True
+            lenv.entries[EncodedString(u"__class__")] = class_node.target.entry
```

Comment:
It shouldn't be necessary to use an `EncodedString` here – unless you use the keys for building the locals dict, as your change does above. Or did you stumble over other things that broke along the way?

Reply:
See above comment -- I think the alternative would be to create an `Entry` with `name=__class__`. I'll try to do it that way.

Reply:
Yes, there should be a new `Entry`. You can use `env.declare_var(name, type, pos)` to create one, like `NameNode.analyse_target_declaration()` does.

Reply:
OK, I've got as far as creating a new `Entry` for `__class__`, but then it fails with local variable `__class__` referenced before assignment. I can also set the name to point to `class_node.target.entry.name`, but that still fails with a similar error. I will keep digging - there doesn't seem to be an obvious way to make these entries synonymous.